zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,329 @@
+    let branch_id = ConsensusBranchId::current(network, height)
+        .expect("must have branch ID for chain history network upgrades");
+    let block_hash = block.hash().0;
+    let time: u32 = block
+        .header
+        .time
+        .timestamp()
+        .try_into()
+        .expect("deserialized and generated timestamps are u32 values");
+    let target = block.header.difficulty_threshold.0;
+    let sapling_root: [u8; 32] = sapling_root.into();
+    let work = block
+        .header
+        .difficulty_threshold
+        .to_work()
+        .expect("work must be valid during contextual verification");
+    // There is no direct `std::primitive::u128` to `bigint::U256` conversion
+    let work = bigint::U256::from_big_endian(&work.as_u128().to_be_bytes());
+
+    let sapling_tx_count = count_sapling_transactions(block);
+
+    zcash_history::NodeData {
+        consensus_branch_id: branch_id.into(),
+        subtree_commitment: block_hash,
+        start_time: time,
+        end_time: time,
+        start_target: target,
+        end_target: target,
+        start_sapling_root: sapling_root,
+        end_sapling_root: sapling_root,
+        subtree_total_work: work,
+        start_height: height.0 as u64,
+        end_height: height.0 as u64,
+        sapling_tx: sapling_tx_count,
+    }
+}
+
+/// Count how many Sapling transactions exist in a block,
+/// i.e. transactions "where either of vSpendsSapling or vOutputsSapling is non-empty"
+/// (https://zips.z.cash/zip-0221#tree-node-specification).
+fn count_sapling_transactions(block: Arc<Block>) -> u64 {
+    block
+        .transactions
+        .iter()
+        .filter(|tx| tx.has_sapling_shielded_data())
+        .count()
+        .try_into()
+        .expect("number of transactions must fit u64")
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{
+        block::Commitment::{self, ChainHistoryActivationReserved},
+        serialization::ZcashDeserializeInto,
+    };
+
+    use super::*;
+    use color_eyre::eyre;
+    use eyre::Result;
+    use zebra_test::vectors::{
+        MAINNET_BLOCKS, MAINNET_FINAL_SAPLING_ROOTS, TESTNET_BLOCKS, TESTNET_FINAL_SAPLING_ROOTS,
+    };
+
+    #[test]
+    fn tree() -> Result<()> {
+        tree_for_network(Network::Mainnet, 903_000)?;
+        tree_for_network(Network::Testnet, 903_800)?;
+        Ok(())
+    }
+
+    fn tree_for_network(network: Network, height: u32) -> Result<()> {
+        let (blocks, sapling_roots) = match network {
+            Network::Mainnet => (&*MAINNET_BLOCKS, &*MAINNET_FINAL_SAPLING_ROOTS),
+            Network::Testnet => (&*TESTNET_BLOCKS, &*TESTNET_FINAL_SAPLING_ROOTS),
+        };
+
+        // Load Block 0 (Heartwood activation block)
+        let block0 = Arc::new(
+            blocks
+                .get(&height)
+                .expect("test vector exists")
+                .zcash_deserialize_into::<Block>()
+                .expect("block is structurally valid"),
+        );
```

Comment:
Type inference, generics, and existing `ZcashDeserialize` impls should get you an `Arc<Block>` using this code:
```suggestion
        // Load Block 0 (network upgrade activation block)
        let block0 = blocks
                .get(&height)
                .expect("test vector exists")
                .zcash_deserialize_into();
```

Reply:
That results in `cannot infer type` :thinking: 

I got that snippet from `vectors.rs`

Reply:
Oh well, it was worth a try 🤷 

We could add utility functions to make this easier, but that's not a high priority right now.